### PR TITLE
Plane: option to immediately climb in AUTO mode (not doing glide slope)

### PR DESCRIPTION
**Pull request by eppravitra**
"Aviation style" climb profile where the aircraft reaches final altitude way before reaching the desired location. 

This has been suggested before in #16248, and I finally got around to giving it a try.

### CODE REVIEW

**Comment by robustini** [comment]
Thanks for the implementation attempt but it doesn't work, tested on SITL and it keeps making direct lines between waypoints.

**Comment by robustini** [comment]
A log demonstrating its successful operation.

[00000026.ZIP](https://github.com/user-attachments/files/16118509/00000026.ZIP)

**Comment by eppravitra** [comment]
I forgot to mention that I only did it for the climb.

Descent still works the same way. Descents will always do glide slope. 

The change works regardless of GLIDE_SLOPE_MIN

To enable the behavior, set FLIGHT_OPTIONS bit 14 to 1

**Comment by shancock884** [comment]
> To enable the behavior, set FLIGHT_OPTIONS bit 14 to 1

If you add a bit to FLIGHT_OPTIONS, remember to add some helpful info about it to Parameters.cpp for the wiki!
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/Parameters.cpp#L1088

**Comment by eppravitra** [comment]
Thanks for the suggestion @shancock884 .

The comment has been added.

**Comment by peterbarker** [comment]
I've marked this for DevCallTopic, and NeedsTesting so it might get tested this weekend.

**Comment by peterbarker** [comment]
@eppravitra why is the parameter `GLIDE_SLOPE_MIN` not sufficient for your purposes?

**Comment by eppravitra** [comment]
Hi @peterbarker ,

Honestly, I totally missed GLIDE_SLOPE_MIN part of the code.

After playing with it, I still stand by my idea that there should be a way to do glide slope during descent and not doing glide slope during climb. 

Setting GLIDE_SLOPE_MIN to 0 effectively disables glide slope for both climb and descent.

I welcome suggestions for better implementation or a better way to organize the parameters though.

**Comment by robustini** [comment]
> Setting GLIDE_SLOPE_MIN to 0 effectively disables glide slope for both climb and descent.

Which is what I had pointed out to you earlier.

**Comment by peterbarker** [comment]
> Honestly, I totally missed GLIDE_SLOPE_MIN part of the code.

Yep, this was also an unfamiliar parameter to me.  tridge knew it existed, 'though!

> After playing with it, I still stand by my idea that there should be a way to do glide slope during descent and not doing glide slope during climb.

I agree.  I think the preference would be to re-use the existing parameter somehow, rather than adding another parameter.

> I welcome suggestions for better implementation or a better way to organize the parameters though.

Two ideas spring to mind.  Negative values mean use for climb only.  Add a `FLIGHT_OPTION` bit to limit it to climb only.

**Comment by eppravitra** [comment]
> Two ideas spring to mind. Negative values mean use for climb only. Add a `FLIGHT_OPTION` bit to limit it to climb only.

Thanks @peterbarker 

I’ll leave it up to you and @tridge to make a call on what is the best way to enable/disable this feature. Just tell me what you guys prefer, and I’m happy to recode.

I’ll admit I do find this a bit confusing. For me, the way I have it (using FLIGHT_OPTIONS) is the least confusing way I can think of.

**Comment by IamPete1** [comment]
We have two bits of mission storage we could use on normal way-points to make this configurable in the mission. Or add a "DO_SET" waypoint for it.

**Comment by timtuxworth** [comment]
Will this work in GUIDED mode? (or if not .. could it?)

**Comment by tridge** [comment]
I'd be happy with a FLIGHT_OPTIONS bit to disable slope climb

File: ArduPlane/Parameters.cpp
Comment (Hwurzburg) [inline ArduPlane/Parameters.cpp:1089]: this description needs work....most dont know "slope" in regards to AUTO

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: ArduPlane/altitude.cpp
Comment (IamPete1) [inline ArduPlane/altitude.cpp:86]:
```suggestion
        if (!above_location_current(next_WP_loc) && plane.flight_option_enabled(FlightOptions::IMMEDIATE_CLIMB_IN_AUTO)) {
```

**Comment by tridge** [comment]
@eppravitra what testing have you done? do you have a SITL log?

**Comment by tridge** [comment]
@robustini did you test this PR?

File: ArduPlane/Parameters.cpp
Comment (eppravitra) [reply ArduPlane/Parameters.cpp]: I added more description as you suggested

**Comment by eppravitra** [comment]
> Will this work in GUIDED mode? (or if not .. could it?)

GUIDED already climbs with maximum TECS performance and descends with slope regardless of this PR.

This PR only affects AUTO mode, effectively allowing AUTO mode to behave the same way as what GUIDED is already doing.

**Comment by eppravitra** [comment]
[log_140_2024-7-16-10-59-17.zip](https://github.com/user-attachments/files/16242986/log_140_2024-7-16-10-59-17.zip)


@tridge 
I just ran another SITL. Please see the attachment.

**Comment by robustini** [comment]
> @robustini did you test this PR in its current form?

@tridge yes and it works as it should, it acts in AUTO obviously only on the climb.
Here a recent log based on the last commit of this PR, with various tests enabling and disabling it.
In my opinion it works perfectly, merge can be done and I am closing my related request about this.

[00000029.ZIP](https://github.com/user-attachments/files/16243291/00000029.ZIP)

**Comment by eppravitra** [comment]
>  Obviously a further implementation that also goes to remove the limit imposed by the GLIDE_SLOPE_MIN without having to set it to 0 would be appreciated but still not essential, it is okay as it is.

This code actually does not require you to change GLIDE_SLOPE_MIN to 0. Just enable the flag and it should work

**Comment by robustini** [comment]
> > Obviously a further implementation that also goes to remove the limit imposed by the GLIDE_SLOPE_MIN without having to set it to 0 would be appreciated but still not essential, it is okay as it is.
> 
> This code actually does not require you to change GLIDE_SLOPE_MIN to 0. Just enable the flag and it should work

I don't see this behavior, you can see it from my log.
To speed up the descent I had to set it to 0.

**Comment by eppravitra** [comment]
Oh, I guess we are talking about different things.

My change only concerns about climb.

File: ArduPlane/Parameters.cpp
Comment (IamPete1) [inline ArduPlane/Parameters.cpp:1089]: I don't think you can use commas here. It confuses the metadata parser.

Comment (eppravitra) [reply ArduPlane/Parameters.cpp]: The comment has been changed

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
per my previous review comment...the label metadata is too long. Something closer to my suggestion is needed..I dont approve still

**Comment by tridge** [comment]
@eppravitra the only thing blocking this PR is the param documentation, the description of the option. Can you please change it to the suggestion from @Hwurzburg (Henry is docs lead)
this PR is ok to merge once that param doc change is made

File: ArduPlane/Parameters.cpp
Comment (eppravitra) [reply ArduPlane/Parameters.cpp]: I made the changes as suggested

**Review by Hwurzburg — APPROVED** [review]

File: ArduPlane/Parameters.cpp
Comment (peterbarker) [inline ArduPlane/Parameters.cpp:1089]:
```suggestion
    // @Bitmask: 14: In AUTO - climb to next waypoint altitude immediately instead of linear climb
```
Can't have commas here.

commas separate bitmask entries, so you can have `Bitmask: 5:Bob, 6:Mary` - a syntax we use elsewhere.  Splitting the bitmask across lines kind of confuses things...

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by eppravitra** [comment]
Thanks!

**Pull request merged**
